### PR TITLE
Map Redesign [Merged into #442 Navigation Fixes]

### PR DESCRIPTION
**Pull request by manunam2002**
### Description

This PR reviews the map markers design, introduces an indicator when the markers are loading, and adds navigation from the ProviderMap markers to the booking screen. It adds documentation for all the Map related functions.

### Changes Made
- **MapMarker**: Redesigned the map marker adhering to the project style.
- **Navigation**: Added navigation to Booking Screen from Provider Map Markers.
- **Chat Assistant**: fixed the prompt to specify who created the service request.

### CODE REVIEW

**Review by yshdb5 — APPROVED** [review]
### Overall Comment  
This PR brings notable improvements to the UI by redoing the map markers, adding a loading indicator, and enabling navigation from ProviderMap markers to the booking screen. The added documentation for map-related functions is a valuable enhancement, supporting clarity and maintainability.

### Strengths of this PR  
- **Loading Indicator**: The loading indicator is a thoughtful addition, improving usability by clearly showing data is being fetched.  
- **Navigation Integration**: Seamlessly connecting map markers to the booking screen simplifies the user journey and improves navigation.  
- **Documentation**: Comprehensive documentation for map-related functions strengthens the maintainability of the codebase.

### Suggestions for Improvement  
- **UI**: I am unsure about how I feel about the design of the maps' markers.

For example, would it be possible to make the markers smaller and get rid of the little icon on the bottom right? It seems a bit too much and in my opinion it doesn't blend in well, what do you think? We can get rid of them or juste find something else that would be better looking?

Capture d’écran 2024-12-12 à 18 57 27

Concerning the loading animation, it's great, but would we make it look more modern so that it blends more within the overall design of the app? 

Capture d’écran 2024-12-12 à 18 59 08

### Conclusion  
This PR effectively enhances the user experience and maintainability of the project. The usability improvements, and navigation integration are strong contributions. Addressing the design issues will further refine the implementation. Great job!

**Comment by manunam2002** [comment]
### Comment on the UI choices

The little icon on the bottom right as I said will reflect the Overview service types Icon, I really don't think it should be removed: If as a seeker I want to find a provider nearby on the map, I surely have a type of service in mind, so how would I know how to click on the right provider? An indication of the provider's service type is probably the most important thing to display on the map.
Regarding the markers size it's a tradeoff between readability and size, smaller than this I find it not readable enough. Either we don't display the name, either we keep it quite big.

File: app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt
Comment (achillet3) [inline app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt:1]: ✅ Cleaner marker design with dedicated icon instead of text snippets 
✅ Smart placeholder approach while waiting for service icons
✅ Added clear loading indicator for markers 
✅ Good user feedback during data fetching
✅ Comprehensive documentation 
✅ Clear separation of concerns between seeker and provider maps

#### Suggestions
1. Map Interaction
```
// Add zoom controls and current location button
@Composable
fun MapControls(
    onZoomIn: () -> Unit,
    onZoomOut: () -> Unit,
    onMyLocation: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier.padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        MapControlButton(
            icon = Icons.Default.Add,
            onClick = onZoomIn
        )
        MapControlButton(
            icon = Icons.Default.Remove,
            onClick = onZoomOut
        )
        MapControlButton(
            icon = Icons.Default.MyLocation,
            onClick = onMyLocation
        )
    }
}
```

2. Service Icon Integration
```
// Prepare for the upcoming service icons
private fun getMarkerIcon(service: Service): Int {
    return Services.getIcon(service) ?: R.drawable.orders_ovw_image // Fallback icon
}
```

File: app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt [hunk 1]
```diff
@@ -130,39 +184,50 @@ fun MapMarker(markerData: MarkerData) {
...
+        OutlinedCard(
+            modifier = Modifier.width(80.dp).height(100.dp).testTag(markerData.tag),
+            shape = RoundedCornerShape(8.dp),
+            elevation = CardDefaults.outlinedCardElevation()) {
+              Box {
+                Image(
+                    bitmap = markerData.image ?: ImageBitmap.imageResource(R.drawable.no_photo),
+                    contentDescription = null,
+                    contentScale = ContentScale.Crop,
+                    modifier = Modifier.fillMaxSize().clip(RoundedCornerShape(8.dp)))
```
Comment (achillet3) [inline app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt:196]:
```suggestion
Box(modifier = Modifier.fillMaxSize()) {
    Image(
        bitmap = markerData.image ?: ImageBitmap.imageResource(R.drawable.no_photo),
        contentDescription = markerData.title,
        contentScale = ContentScale.Crop,
        modifier = Modifier
            .fillMaxSize()
            .clip(RoundedCornerShape(8.dp))
            .testTag("${markerData.tag}_image")
    )
    if (markerData.image == null) {
        Icon(
            painter = painterResource(R.drawable.ic_error),
            contentDescription = "Image load error",
            modifier = Modifier
                .align(Alignment.TopEnd)
                .padding(4.dp)
                .size(16.dp),
            tint = colorScheme.error
        )
    }
}
```

File: app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt [hunk 2]
```diff
@@ -121,6 +170,11 @@ fun MapContent(
...
 }
 
+/**
+ * Composable that displays a marker on the map.
+ *
+ * @param markerData The data to display on the marker.
+ */
 @Composable
 fun MapMarker(markerData: MarkerData) {
   MarkerComposable(
```
Comment (achillet3) [inline app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt:180]: You could animate the selected markers with a code of this style
```
val selected = remember { mutableStateOf(false) }
MarkerComposable(
    keys = arrayOf(markerData.tag),
    state = MarkerState(position = markerData.location),
    onClick = {
        selected.value = !selected.value
        markerData.onClick()
        true
    }
) {
    OutlinedCard(
        modifier = Modifier
            .width(80.dp)
            .height(100.dp)
            .testTag(markerData.tag)
            .scale(animateFloatAsState(if (selected.value) 1.1f else 1f).value),
        shape = RoundedCornerShape(8.dp),
        elevation = CardDefaults.outlinedCardElevation(
            defaultElevation = if (selected.value) 8.dp else 4.dp
        )
    ) {
        // ... existing content ...
    }
}
```

File: app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt [hunk 3]
```diff
@@ -80,13 +104,38 @@ fun MapScreen(userLocation: LatLng?, markers: List<MarkerData>, bottomBar: @Comp
...
+                    elevation = CardDefaults.outlinedCardElevation()) {
+                      Row(
+                          modifier = Modifier.padding(8.dp),
+                          verticalAlignment = Alignment.CenterVertically) {
+                            Text("Loading images  ", fontWeight = FontWeight.Bold)
+                            CircularProgressIndicator(
+                                modifier = Modifier.size(30.dp), color = colorScheme.onBackground)
+                          }
+                    }
+              }
```
Comment (achillet3) [inline app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt:122]: You could animate the loading a little bit more with an Animated Visibility
```
AnimatedVisibility(
    visible = markersLoading,
    enter = fadeIn() + slideInVertically(),
    exit = fadeOut() + slideOutVertically()
) {
    Column(
        modifier = Modifier
            .align(Alignment.TopCenter)
            .padding(16.dp)
            .testTag("loading_indicator")
    ) {
        OutlinedCard(
            shape = RoundedCornerShape(32.dp),
            elevation = CardDefaults.outlinedCardElevation()
        ) {
            Row(
                modifier = Modifier.padding(8.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text(
                    "Loading images",
                    fontWeight = FontWeight.Bold,
                    style = MaterialTheme.typography.labelLarge
                )
                CircularProgressIndicator(
                    modifier = Modifier.size(24.dp),
                    color = colorScheme.primary,
                    strokeWidth = 2.dp
                )
            }
        }
    }
}
```

**Review by achillet3 — APPROVED** [review]
# Review of PR #446: Map Redesign 🗺️

## Overview
This PR introduces significant improvements to the map functionality, focusing on marker design, loading states, and navigation. The changes demonstrate good attention to UI/UX details and code organization.

## 🌟 Strengths

### 1. UI Implementation
- Excellent use of Material 3 components (`OutlinedCard`, `CircularProgressIndicator`)
- Clean and consistent styling with proper padding and alignment
- Well-implemented loading indicator that provides clear user feedback
- Proper use of Material Design elevation and shapes

### 2. Code Organization
- Clear separation of concerns in map-related components
- Well-structured composable functions with single responsibilities
- Consistent use of modifiers and styling parameters
- Good state management with `markersLoading` flag

### 3. Documentation
- Comprehensive KDoc comments for all public functions
- Clear parameter descriptions
- Well-documented composable functions
- Consistent documentation style throughout the changes

### 4. Navigation
- Clean implementation of marker click handling
- Direct navigation to booking details from provider map
- Proper state management before navigation

## 🚨 Concerns

### 1. Error Handling
- Limited visual feedback for failed image loads
- No retry mechanism for failed resource loading
- Missing error logging for debugging purposes

### 2. Testing
- Basic test coverage could be expanded
- Missing edge case tests (network failures, null states)
- No performance tests for large marker sets

## 🔧 Suggested Improvements

### 1. Enhanced Error Visualization
- Add visual indicator for failed image loads
- Implement error state in marker UI
- Show error messages when appropriate
- Add accessibility descriptions for error states

### 2. Loading State Polish
- Add smooth animations for loading state transitions
- Implement staggered loading for markers
- Enhance loading indicator with progress information
- Consider skeleton UI for loading markers

### 3. Additional Test Coverage
- Add tests for loading states
- Implement error state testing
- Add performance benchmarks
- Include accessibility testing

## 🎯 Future Considerations

1. **Performance Optimization**
   - Consider implementing marker clustering for large datasets
   - Add viewport-based marker loading
   - Implement image caching strategy

2. **User Experience**
   - Add marker animations for selection/deselection
   - Consider adding map controls (zoom, center on user)
   - Implement marker filtering options

3. **Error Recovery**
   - Add retry mechanism for failed image loads
   - Implement proper error logging
   - Add user feedback for network issues

## 📝 Overall Recommendation

**Approve with enhancement suggestions** ✅

The PR demonstrates solid implementation with good attention to UI/UX details. The suggested improvements would add extra polish to an already well-structured implementation. The core functionality works as expected, and the code organization follows best practices.

### Priority of Improvements
1. 🔴 High: Enhanced error visualization and logging
2. 🟡 Medium: Additional test coverage
3. 🟢 Low: Animation and UX enhancements

These improvements can be implemented in follow-up PRs if needed, as they don't affect the core functionality of the current implementation.

File: app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt
Comment (manunam2002) [reply app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt:1]: We already have the zoom MapControls!

File: app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt (hunk 2 above)
Comment (manunam2002) [reply app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt:180]: since we navigate onClick this is not useful

File: app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt (hunk 3 above)
Comment (manunam2002) [reply app/src/main/java/com/android/solvit/shared/ui/map/MapScreen.kt:122]: Actually `AnimatedVisibility` cannot be used inside a `Box` composable, what a shame! this would have been very cool
